modules/standard/Maps.chpl

```diff
@@ -0,0 +1,291 @@
+    */
+    proc init=(const ref other: map(?kt, ?vt), param parSafe=other.parSafe) {
+
+    }
+
+    /*
+      Clears the contents of this map.
+
+      .. warning::
+
+        Clearing the contents of this map will invalidate all existing
+        references to the elements contained in this map.
+    */
+    proc clear() {
+
+    }
+
+    /*
+      The current number of keys contained in this map.
+    */
+    inline proc const size {
+
+    }
+
+    /*
+      Returns `true` if this map contains zero keys.
+
+      :returns: `true` if this map is empty.
+      :rtype: `bool`
+    */
+    inline proc const isEmpty(): bool {
+
+    }
+
+    /*
+      Returns `true` if the given key is a member of this map, and `false`
+      otherwise.
+
+      :arg k: The key to test for membership.
+      :type k: keyType
+
+      :returns: Whether or not the given key is a member of this map.
+      :rtype: `bool`
+    */
+    proc const contains(const ref k: keyType): bool {
+
+    }
+
+    /*
+      Updates this map with the contents of the other, overwritting the values
+      for already-existing keys.
+
+      :arg m: The other map
+      :type m: map(keyType, valType)
+    */
+    proc update(const ref m: zap(keyType,valType)) {
+
+    }
+
+    /*
+      Get a reference to the value mapped to the given key.
+
+      :arg k: The key to access
+      :type k: keyType
+
+      .. warning::
+
+        Use of the `this` method with a non-existent key (while bounds checking
+        is on) will cause the currently running program to halt.
+
+      :returns: Reference to the value mapped to the given key.
+    */
+    proc const this(k: keyType) ref {
+
+    }
+
+    /*
+      Iterates over the keys of this map.
+
+      :yields: A reference to one of the keys contained in this map.
+    */
+    iter these() const ref {
+
+    }
+
+    /*
+      Iterates over the key-value pairs of this map.
+
+      :yields: A tuple of references to one of the key-value pairs contained in
+               this map.
+    */
+    iter items() const ref {
+
+    }
+
+    /*
+      Iterates over the values of this map.
+
+      :yields: A reference to one of the values contained in this map.
+    */
+    iter values() ref {
+
+    }
+
+    /*
+      Writes the contents of this map to a channel. The format looks like:
+
+        .. code-block:: chapel
+    
+           {k1: v1, k2: v2, .... , kn: vn}
+
+      :arg ch: A channel to write to.
+    */
+    proc writeThis(ch: channel) {
+
+    }
+
+    /*
+      Adds a key-value pair to the map. Different than doing `myMap[key] =
+      value`, this method returns `false` if the key already exists in the map.
+
+     :arg k: The key to add to the map
+     :type k: keyType
+
+     :arg v: The value that maps to ``k``
+     :type k: valueType
+
+     :returns: `true` if `k` was not in the map and added with value `v`.
+               `false` otherwise.
+     :rtype: bool
+    */
+    proc add(k: keyType, v: valType): bool {
```

Comment:
Is there a particularly strong motivator for having the semantics of this be "add if not present" over just a more general purpose "add(key, value)" or "put(key, value)"?

Reply:
Not a strong one. `putIfAbsent` in Java sounded like a good idea as it can allow nicer idioms. But I wasn't quite fond of the long name, and `put` doesn't really convey the message that this method is supposed to add something to the data structure.

Reply:
I like `set` better than `put`. I think we should avoid `put` since we are a PGAS language.

Reply:
I think this seems to be the most important point we need to settle on.

- Is there any objection to this method returning `false` if the key already exists in the map. If it adds/sets the value even if the key exists, then, it'll be equivalent to `myMap['foo'] = 'bar'`

- Good point on `put` being a PGAS term. If we are keeping the semantics as-is, I'd still like this method to be called `add`. At the same time we can also have a `set` which sets the value associated with a key only if the key already exists in the map. I think such an `add` and `set` pair complement each other well. Any objections to that?